src/EditorFeatures/Core/Tagging/AsynchronousViewportTaggerProvider.SingleViewportTaggerProvider.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+internal abstract partial class AsynchronousViewportTaggerProvider<TTag> where TTag : ITag
+{
+    /// <summary>
+    /// Actual <see cref="AsynchronousViewTaggerProvider{TTag}"/> responsible for tagging a particular span (or spans)
+    /// of the view.  Inherits all behavior of a normal view tagger, except for determining what spans to tag and what
+    /// cadence to tag them at.
+    /// </summary>
+    private sealed class SingleViewportTaggerProvider : AsynchronousViewTaggerProvider<TTag>
+    {
+        private readonly AsynchronousViewportTaggerProvider<TTag> _callback;
+
+        private readonly ViewPortToTag _viewPortToTag;
+
+        public SingleViewportTaggerProvider(
+            AsynchronousViewportTaggerProvider<TTag> callback,
+            ViewPortToTag viewPortToTag,
+            IThreadingContext threadingContext,
+            IGlobalOptionService globalOptions,
+            ITextBufferVisibilityTracker? visibilityTracker,
+            IAsynchronousOperationListener asyncListener)
+            : base(threadingContext, globalOptions, visibilityTracker, asyncListener)
+        {
+            _callback = callback;
+            _viewPortToTag = viewPortToTag;
+        }
+
+        protected override ImmutableArray<IOption2> Options
```

Comment:
bridges from the actual individual *view* taggers, to the outer *viewport* tagger which allows subclasses to customize behavior.